### PR TITLE
TrendCarouselITest

### PR DESCRIPTION
**Pull request by RedTo**
Integration test for trend carousel, with new page objects.

### CODE REVIEW

File: src/test/java/io/jenkins/plugins/analysis/warnings/recorder/TrendCarouselITest.java [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+import hudson.model.Result;
+
+import io.jenkins.plugins.analysis.core.model.AnalysisResult;
+import io.jenkins.plugins.analysis.core.steps.IssuesRecorder;
+import io.jenkins.plugins.analysis.core.testutil.IntegrationTestWithJenkinsPerSuite;
+import io.jenkins.plugins.analysis.warnings.Java;
+import io.jenkins.plugins.analysis.warnings.recorder.pageobj.TrendCarousel;
+import io.jenkins.plugins.analysis.warnings.recorder.pageobj.TrendCarousel.TrendChartType;
+
+import static org.assertj.core.api.Assertions.*;
```
[the remark this answers is not in the document]
Comment (RedTo) [reply src/test/java/io/jenkins/plugins/analysis/warnings/recorder/TrendCarouselITest.java]: This one is necessary!

[the remark this answers is not in the document]
Comment (RedTo) [reply src/test/java/io/jenkins/plugins/analysis/warnings/recorder/TrendCarouselITest.java]: This one is necessary!

[the remark this answers is not in the document]
Comment (uhafner) [reply src/test/java/io/jenkins/plugins/analysis/warnings/recorder/TrendCarouselITest.java]: I think you need to merge/rebase with master. I updated the rules some time ago.

File: src/test/java/io/jenkins/plugins/analysis/warnings/recorder/TrendCarouselITest.java [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+    }
+
+    /**
+     * Test that tools trend chart is next.
+     */
+    @Test
+    public void shouldShowToolsTrendChartAsNext() {
+        TrendCarousel carousel = setUpTrendChartTest(false);
+
+        assertThat(carousel.next().equals(carousel.getActive())).isTrue();
```
Comment (uhafner) [inline src/test/java/io/jenkins/plugins/analysis/warnings/recorder/TrendCarouselITest.java]: `assertThat(carousel.next()).isEqualTo(carousel.getActive());`

File: src/test/java/io/jenkins/plugins/analysis/warnings/recorder/TrendCarouselITest.java [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+ *
+ * @author Tobias Redl
+ * @author Andreas Neumeier
+ */
+
+public class TrendCarouselITest extends IntegrationTestWithJenkinsPerSuite {
+
+    private HtmlPage webPage;
+
+    /**
```
Comment (uhafner) [inline src/test/java/io/jenkins/plugins/analysis/warnings/recorder/TrendCarouselITest.java]: In unit tests a setup with very small tests would be perfect. In integration tests it makes more sense to combine as many variants as possible in order to speed up the tests. I basically see only one or two big scenarios:
start -> next -> next -> next -> next -> prev -> prev -> prev -> prev -> end and the additional tests for health and local storage.

File: src/test/java/io/jenkins/plugins/analysis/warnings/recorder/TrendCarouselITest.java [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+     * Test that tools trend chart is shown after three clicks on next.
+     */
+    @Test
+    public void shouldShowSeveritiesTrendChartAfterThreeTimesNext() {
+        TrendCarousel carousel = setUpTrendChartTest(false);
+
+        carousel.next(3);
+        assertThat(carousel.getActiveChartType()).isEqualTo(TrendChartType.SEVERITIES);
```
Comment (uhafner) [inline src/test/java/io/jenkins/plugins/analysis/warnings/recorder/TrendCarouselITest.java]: These tests actually test the page object and not the UI.

File: src/test/java/io/jenkins/plugins/analysis/warnings/recorder/TrendCarouselITest.java [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+            throw new AssertionError("WebPage refresh failed.", e);
+        }
+        assertThat(carousel.getActiveChartType()).isEqualTo(TrendChartType.TOOLS);
+    }
+
+    /**
+     * Tests that only one of all three charts is shown as active.
+     */
+    @Test
+    public void shouldShowOnlyOneOfThreeChartsAsActive() {
```
Comment (uhafner) [inline src/test/java/io/jenkins/plugins/analysis/warnings/recorder/TrendCarouselITest.java]: This is the same test as `shouldShowSeveritiesTrendChartAsDefault`just with more assertions.

File: src/test/java/io/jenkins/plugins/analysis/warnings/recorder/TrendCarouselITest.java [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+        TrendCarousel carousel = setUpTrendChartTest(false);
+
+        assertThat(carousel.getActiveChartType()).isEqualTo(TrendChartType.SEVERITIES);
+        assertThat(carousel.getActiveChartType()).isNotEqualTo(TrendChartType.TOOLS);
```
Comment (uhafner) [inline src/test/java/io/jenkins/plugins/analysis/warnings/recorder/TrendCarouselITest.java]: This assertion actually tests the enum implementation of Java (or the `isNotEqualTo` method assertJ).

File: src/test/java/io/jenkins/plugins/analysis/warnings/recorder/TrendCarouselITest.java (hunk 4 above)
Comment (RedTo) [reply src/test/java/io/jenkins/plugins/analysis/warnings/recorder/TrendCarouselITest.java]: @uhafner Do you mean that the tests were to small, or that our assertions are not as they should be?

Comment (uhafner) [reply src/test/java/io/jenkins/plugins/analysis/warnings/recorder/TrendCarouselITest.java]: No, the comment actually was about the functionality next(2) and next(3). These methods are not in the actual UI, they are a functionality of the page object. 

So navigating step by step and validating each step is the best approach (as you are doing now).

**Pull request merged**